Add WithUsingResource to generic AST

Sjord opened the pull request:
Obtain a resource using some expression or variable declaration, and dispose of
that value at the end of the block.

The first statement would be a ExprStmt or DefStmt to acquire some resource,
and the second statement would be the block where the resource is used.

E.g. in python:
```python
with fopen(file) as fp:
	...
```

E.g. in C#:
```csharp
using (var reader = new StringReader(manyLines))
{ ... }
```

E.g. in Java:
```java
try (Foo f = new Foo())
{ ... }
```

Resolves #342

Sjord committed:
Add WithUsingResource to generic AST

Obtain a resource using some expression or variable declaration, and dispose of
that value at the end of the block.

The first statement would be a ExprStmt or DefStmt to acquire some resource,
and the second statement would be the block where the resource is used.

E.g. in python:
```python
with fopen(file) as fp:
	...
```

E.g. in C#:
```csharp
using (var reader = new StringReader(manyLines))
{ ... }
```

E.g. in Java:
```java
try (Foo f = new Foo())
{ ... }
```

Resolves #342

Sjord commented:
Should this have `(* newvar: *)` or `(* newscope: *)` comments somewhere?

aryx commented:
newscope: yes, newvar: no.
(note that those comments are not important; they are informal hints for people writing program analysis
using the generic AST).

Sjord committed:
Add newscope: comment to block of WithUsingResource

The body of a using-statement is a new scope.

aryx commented:
So I don't think you actually build pfff with this construct, because if you do so (with dune build),
you should get many errors in MAP_ast.ml, Visitor_AST.ml, where you have to implement the code
for this new construct.
Let me know if it's too complicated; I can merge your code and fix it myself later.

aryx commented:
It's mostly boilerplate code to fill for visitors, dumpers, and mappers of the generic AST.
There will be also repercursions in the semgrep repository, but we can do that in a separate PR (well we have to anyway since it's a different repo).

aryx commented:
hopefully the CI will fail and tell you where to fix the code; but really you should try 'dune build' instead of waiting each time for the CI to display a new error.

Sjord committed:
Add WithUsingResource

For C#'s using statement, or Java's try-with-resource.

Treat it as two consecutive statements in most places.

aryx commented:
hmm, you may need to rebase with the latest develop because martin ran a huge script to reindent and fix trailing space in the pfff codebase.
Sorry for the inconvenience

Sjord committed:
Fix indention
